plutus-benchmark/flat/Main.hs

```diff
@@ -0,0 +1,86 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Main where
+
+import           Control.DeepSeq              (force)
+import           Criterion.Main
+import           Criterion.Types
+import qualified Data.ByteString              as BS
+import           Data.Map.Strict              (Map, empty, insert, lookup)
+import           Data.Maybe                   (fromJust)
+import           Data.Text                    (Text, unpack)
+import           System.Mem                   (performMajorGC)
+
+import           Language.PlutusCore          (Name (..))
+import           Language.PlutusCore.Universe
+import           Language.UntypedPlutusCore
+
+import           Codec
+import           Dataset
+import           Report
+
+import           Prelude                      hiding (lookup)
+
+type Tm = Term Name DefaultUni ()
+
+serializedContracts :: Map (Text, Text) BS.ByteString
+serializedContracts =
+  foldl update empty [ (contract, codec) | contract <- contracts, codec <- codecs ]
+  where
+    update :: Map (Text, Text) BS.ByteString
+           -> ((Text, Tm), (Text, Codec Tm))
+           -> Map (Text, Text) BS.ByteString
+    update cs ((contractName, contract), (codecName, codec)) =
+      insert (contractName, codecName) ((Codec.serialize codec) contract) cs
+
+benchmarkSerialize ::
+     (Text, Tm)
+  -> (Text, Codec Tm)
+  -> Benchmark
+benchmarkSerialize (contractName, contract) (codecName, (Codec serialize' _)) =
+  bench (unpack $ " " <> contractName <> " - " <> codecName)
+        -- (nf (BS.length . serialize) contract) ~ original code
+        (nf serialize' contract)
+
+benchmarkDeserialize ::
+     Map (Text, Text) BS.ByteString
+  -> (Text, Tm)
+  -> (Text, Codec Tm)
+  -> Benchmark
+benchmarkDeserialize scs (contractName, _) (codecName, codec) =
+  let !serialized = fromJust $ lookup (contractName, codecName) scs
+  in  bench (unpack $ " " <> contractName <> " - " <> codecName)
+            -- (whnf ((contract ==) . (Codec.deserialize codec)) serialized) ~ original code
+            (whnf (Codec.deserialize codec) serialized)
+
+benchmarkContract ::
+     Map (Text, Text) BS.ByteString
+  -> (Text, Tm)
+  -> [Benchmark]
+benchmarkContract scs contract =
+  [ bgroup "serialization (time) " $
+       map (benchmarkSerialize contract) codecs
+   , bgroup "deserialization (time) " $
+       map (benchmarkDeserialize scs contract) codecs
+  ]
+
+benchmarkConfig :: Config
+benchmarkConfig = defaultConfig
+  { jsonFile   = Just "flat-report.json"
+  , reportFile = Just "flat-report.html"
+  }
+
+sizesReport :: FilePath
+sizesReport = "flat-sizes.txt"
+
+main :: IO ()
+main = do
+  let !scs = force serializedContracts
+      !contracts' = force contracts
+  performMajorGC
+  defaultMainWith benchmarkConfig $
+    concatMap (benchmarkContract scs) contracts'
+  putStrLn "Measuring sizes.."
```

Comment:
Nitpick: Should this be a full stop or an ellipsis?

Reply:
I modified the message and added an extra `.`.